test/e2e/latency.go

```diff
@@ -0,0 +1,277 @@
+		scheduleToWatchLatencies = make([]podLatencyData, 0)
+		e2eLatencies             = make([]podLatencyData, 0)
+	)
+
+	for name, podNode := range nodes {
+		createTs, ok := createTimestamps[name]
+		Expect(ok).To(Equal(true))
+		scheduleTs, ok := scheduleTimestamps[name]
+		Expect(ok).To(Equal(true))
+		runTs, ok := startTimestamps[name]
+		Expect(ok).To(Equal(true))
+		watchTs, ok := watchTimestamps[name]
+		Expect(ok).To(Equal(true))
+
+		var (
+			scheduleLatency        = podLatencyData{name, podNode, scheduleTs.Time.Sub(createTs.Time)}
+			startLatency           = podLatencyData{name, podNode, runTs.Time.Sub(scheduleTs.Time)}
+			watchLatency           = podLatencyData{name, podNode, watchTs.Time.Sub(runTs.Time)}
+			scheduleToWatchLatency = podLatencyData{name, podNode, watchTs.Time.Sub(scheduleTs.Time)}
+			e2eLatency             = podLatencyData{name, podNode, watchTs.Time.Sub(createTs.Time)}
+		)
+
+		scheduleLatencies = append(scheduleLatencies, scheduleLatency)
+		startLatencies = append(startLatencies, startLatency)
+		watchLatencies = append(watchLatencies, watchLatency)
+		scheduleToWatchLatencies = append(scheduleToWatchLatencies, scheduleToWatchLatency)
+		e2eLatencies = append(e2eLatencies, e2eLatency)
+	}
+
+	sort.Sort(latencySlice(scheduleLatencies))
+	sort.Sort(latencySlice(startLatencies))
+	sort.Sort(latencySlice(watchLatencies))
+	sort.Sort(latencySlice(scheduleToWatchLatencies))
+	sort.Sort(latencySlice(e2eLatencies))
+
+	printLatencies(scheduleLatencies, "worst schedule latencies")
+	printLatencies(startLatencies, "worst run-after-schedule latencies")
+	printLatencies(watchLatencies, "worst watch latencies")
+	printLatencies(scheduleToWatchLatencies, "worst scheduled-to-end total latencies")
+	printLatencies(e2eLatencies, "worst e2e total latencies")
+
+	// Test whether e2e pod startup time is acceptable.
+	// TODO: Switch it to 5 seconds once we are sure our tests are passing.
+	podStartupThreshold := 8 * time.Second
+	e2ePodStartupTime50perc := startLatencies[len(startLatencies)/2].Latency
+	e2ePodStartupTime90perc := startLatencies[len(startLatencies)*9/10].Latency
+	e2ePodStartupTime99perc := startLatencies[len(startLatencies)*99/100].Latency
+	Expect(e2ePodStartupTime50perc).To(BeNumerically("<", podStartupThreshold), "Too high pod startup time 50th percentile")
+	Expect(e2ePodStartupTime90perc).To(BeNumerically("<", podStartupThreshold), "Too high pod startup time 90th percentile")
+	Expect(e2ePodStartupTime99perc).To(BeNumerically("<", podStartupThreshold), "Too high pod startup time 99th percentile")
+
+	// Log suspicious latency metrics/docker errors from all nodes that had slow startup times
+	for _, l := range startLatencies {
+		if l.Latency > NodeStartupThreshold {
+			HighLatencyKubeletOperations(c, 1*time.Second, l.Node)
+		}
+	}
+
+	Logf("Approx throughput: %v pods/min",
+		float64(nodeCount)/(startLatencies[len(startLatencies)-1].Latency.Minutes()))
+}
```

Comment:
There is huge intersection between the code if this test and the code in density. Is it possible to refactor it to avoid code duplication?